tensorflow/compiler/mlir/hlo/include/mlir-hlo/utils/elementwise_op_list.h

```diff
@@ -0,0 +1,21 @@
+#ifdef GET_ELEMENTWISE_OP_LIST
+#undef GET_ELEMENTWISE_OP_LIST
```

Comment:
Why isn't this reusing element wise trait?

Reply:
In my current knowledge of trait, we can only add a interface to tell "if an lhlo is elementwise", but we don't have a method to return "the collection of all the elementwise ops" at compile time? (as these are to be passed to templates). Please correct me if i miss something. And also there's another reason, here it not only requires the lhlo_op to be elementwise, but also requires it to be supported in HloOpToStdScalarOp, which is not the same thing for now.

Reply:
If "the collection of all the elementwise ops" is only because `HloOpToStdScalarOp::map<LHLO_OpTy>` is a C++ template, then I think we should fix `HloOpToStdScalarOp::map<LHLO_OpTy>` first. Then maybe we can use the elementwise trait Jacques proposed.

Reply:
> > Other than the `elemwise{Fuse,Lower}HelperOr` and that we should adopt the existing elementwise trait, I don't have other comments worth mentioning. I do believe that if I take a microscope I'll have more comments, but I decided to leave out those nits.
> 
> Here it is not only because of the existing mapping of lhlo -> std is with templates, but also the elementalLower<> in lhlo_elemental_utils.cc is also implemented with templates. Here "elementwise_op_list.h" should be renamed as "disc_supported_elemwise_op_list.h", will this be better and acceptable? Anyhow we need to maintain a "closed collection of all supported possible ops" for a certain compiler implementation, such kind of information should not be put in the dialects itself.
> 
> BTW I also agree it's not strictly necessary to mapping lhlo -> std with templates. However the fact is map_lmhlo_to_scalar_op.h has already been used in several existing passes (lhlo_legalize_to_affine, legalize_to_linalg & lhlo_legalize_to_gpu). The code amount of map_lmhlo_to_scalar_op.h is more than 1000 lines, so i believe the effort to get rid of templates will not be trivial. Maybe we should hear the comments of Adrian Kuegel [akuegel@google.com](mailto:akuegel@google.com) here (how to @adrian Kuegel here?).

OK, there are several things tangled together, let me try to untangle them:
* I think for general optimization/lowering passes, static typing make sense, as those passes have to access *specific* ODS APIs of each op. These passes run a "visitor pattern" to visit different nodes. `elemwise{Fuse,Lower}HelperOr` fall into this category, and I now think they are fine as they are.
* I don't think that `elementwise_op_list.h` should be implemented in terms of a "visitor pattern". It should be like a trait, because users don't care about specifics of each op.

You are also right that the implementation has to maintain a closed collection of ops, but they are for the first case, not the second case.

Reply:
As discussed offline, I changed elementwise_op_list.h into 'disc_supported_list.h". Please have a look.

Reply:
Please be aware i still need two distinguishing macros in 'disc_supported_list.h' GET_SUPPORTED_ELEMENTWISE_OP_LIST and GET_SUPPORTED_NON_ELEMENTWISE_OP_LIST.  As they has to be template parameters.

Reply:
I see that we are already using `hasTrait<OpTrait::Elementwise>()` in `elemwiseFuseHelper` and `elemwiseLowerHelper` to filter out non-elementwise ops, so why bother with `GET_SUPPORTED_ELEMENTWISE_OP_LIST`, which filters the list again? Wouldn't an overall list of `GET_SUPPORTED_OP_LIST` just work?

Reply:
HloOpToStdScalarOp::map<LHLO_OpTy>() is not able to handle non-elementwise LHLO_OpTy during compilation. I may have two variant ways to do this:
1, Revise HloOpToStdScalarOp with a default implementation for other LHLO_OpTys. I looked into the codes and feels the code change will be tricky and might violate with the purpose of the original contributor.
2, Function template partial specialization with std::enable_if, this is not directly supported and must be with other tricks, like a class wrapper. I don't feel this is necessary.
3, distinguishing GET_SUPPORTED_ELEMENTWISE_OP_LIST and GET_SUPPORTED_NON_ELEMENTWISE_OP_LIST in the supported op list. 

I feel 3 is more simply and intuitive. I may also take a deeper look into 1, but I feel the code change will not be very small and we may need to involve Adrian Kuegel  for code review.

Reply:
As I experimented, `lmhlo::HloOpToStdScalarOp::map` has a default implementation (which simple returns `nullptr`) [1], and it also support non-elementwise ops like `lmhlo::DotOp` [2]. I also tried with a supposedly unsupported op (`mhlo::DotOp`, not `lmhlo`) and it also compiles (which will return `nullptr` at runtime).

[1] https://github.com/tensorflow/mlir-hlo/blob/38496e2a88113c5321a5aa50a56cca01f4727de6/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h#L102
[2] https://github.com/tensorflow/mlir-hlo/blob/38496e2a88113c5321a5aa50a56cca01f4727de6/lib/Dialect/mhlo/transforms/lhlo_legalize_to_affine.cc#L84

Reply:
Updated after picking 081e08c5e032f15c678e6e7e69867b08b81efe0d.